### ISSUE TITLE
Fix issue where multiple Drupal users share the same email.

### ISSUE THREAD

**Issue by DFurnes** [DoSomethingArchive/legacy-website#6409]
#### Current Behavior

There are currently a bunch of users (both old and new) that are registered with the same value in the `mail` column in the users table. This is kinda a bummer since we should really only have one account per email. 😞 📬 
#### Desired Behavior

We need to do some more research on how Drupal is currently treating users with duplicate emails (like when you try to log in with that email, which account does it pick?), and then remove the duplicate users who are not being actively used (and perhaps merge their activity into the "active" accounts?)
#### Steps to Replicate

Try running this real outrageous 🐬 query:

``` sql
SELECT group_concat(u.uid), u.name, u.mail, m.field_mobile_value, s.field_user_registration_source_value, from_unixtime(u.access), count(u.mail)
FROM users u
LEFT JOIN field_data_field_mobile m on m.entity_id = u.uid 
left join field_data_field_user_registration_source s on s.entity_id = u.uid
GROUP BY u.mail
HAVING COUNT(u.mail) > 1
ORDER BY u.mail
```
#### Why This Matters

When trying to sync users with Northstar, we're getting validation errors because we're trying to make multiple users that all share the same "unique" field. It's also super confusing how this even _works_ in Phoenix so it'd be nice to figure out that mystery!

📪 🌵

**Comment by DFurnes** [comment]
New query that seems to be a bit more helpful:

``` sql
SELECT u.uid, u.mail, from_unixtime(u.created), from_unixtime(u.access), m.field_mobile_value, s.field_user_registration_source_value FROM users u
LEFT JOIN field_data_field_mobile m ON m.entity_id = u.uid 
LEFT JOIN field_data_field_user_registration_source s ON s.entity_id = u.uid
WHERE mail IN (SELECT mail FROM users GROUP BY mail HAVING COUNT(mail) > 1);
```

**Comment by mshmsh5000** [comment]
In case we haven't posted the explanation elsewhere:
- There's no `UNIQUE` constraint on `mail` in the `users` table, so MySQL's not gonna stop you.
- Drupal's default user form validation will [catch duplicate email addresses](https://github.com/drupal/drupal/blob/7.x/modules/user/user.module#L1273), so these are most likely from when we've bypassed either that validation or form submission entirely (API calls?).

**Comment by angaither** [comment]
Most (if not all) of the cases appear to be two (or more) users created right in a row. the ol' double ping. the clickly-click. 

@mshmsh5000 can we add `UNIQUE` to mail? would that be the worst?

**Comment by DFurnes** [comment]
@mshmsh5000 Yeah, there was a period where we didn't have validation running on API requests (which was added [in September 2014](https://github.com/DoSomething/phoenix/pull/3034)). That explains all of the dupes that exist on staging, but unfortunately they're still showing up regularly on production. 😢

**Comment by mshmsh5000** [comment]
I think it makes _super hella sense_ to have `UNIQUE` on `mail`. Obv we'd have to eliminate the current dupes first.

If the duplicates are occurring serially like that, I wonder if there's some double-firing of `user_save()` _after_ the form/API validation.

**Comment by DFurnes** [comment]
I'm not sure if this is responsible for _all_ the dupes, but I am able to reliably create duplicate users by triggering two requests right after one another. The ol' race condition. 🐎 

Proof of concept using CURL:

``` sh
curl -X "POST" "http://dev.dosomething.org:8888/api/v1/users" -H "Content-Type: application/json" -d "{\"email\":\"sneaky-dupe@example.com\",\"password\":\"test123\"}" & curl -X "POST" "http://dev.dosomething.org:8888/api/v1/users" -H "Content-Type: application/json" -d "{\"email\":\"sneaky-dupe@example.com\",\"password\":\"test123\"}"
```

**Comment by DFurnes** [comment]
And (as suggested), adding a `UNIQUE` constraint on the `mail` column fixes that up. 👮

**Comment by DFurnes** [comment]
Ok, so the things we need to do (testing first on Thor):
- [x] Make sure you've deployed the latest code, like a reasonable person.
- [x] Remove old login tokens (made before app launch) from Northstar's database: `db.tokens.remove({created_at: {$lt: new Date('05/01/2016')}});`
- [x] Set the socket timeout to be really long, because 30 seconds is TOO QUICK for the important work we are about to embark on. Add this to Northstar's Mongo connection settings:  `'socketTimeoutMS' => 9000000`
- [x] Run the `php artisan northstar:clean_drupal_ids --force` script on Northstar. This will delete any profiles with duplicate Drupal IDs on Northstar (and attempt to keep the "correct" profile based on choosing the one with the earliest creation date).

Next, run the Drush scripts. These will each fix users in Phoenix for each type of problem, and attempt to update the corresponding record in Northstar (or create one if that 404s):
- [x] Run the `sanitize-mobile-numbers` script.
- [x] Run the `unset-duplicate-mobiles` script.
- [x] Run the `remove-duplicate-emails` script.
- [ ] Re-run the `export-users-to-northstar` script as a spot-check that everything is good (and to get the remaining missing mobile users). This will also catch any remaining conflicts, and add them to the delete queue so we can remove them in the next (and final!) step.
- [ ] Run the `delete-bad-users-in-northstar` script to fix up those remaining conflicts that popped up on the export run-through.

**Comment by DFurnes** [comment]
Another issue that occurred to me is that we're not normalizing email addresses on Phoenix, so `Example@example.com` and `example@example.com` would count as two unique values (whereas Northstar normalizes them to both be all lowercase).

We'll need to add a script to fix this and then a form alter hook to normalize new registrations.

**Comment by DFurnes** [comment]
Oh, actually I lied – looks like Drupal has been good about that: I can't seem to register with a duplicate differently cased email address, and running a `GROUP BY LCASE(mail)` query doesn't return any extra duplicates. ✅

**Comment by mshmsh5000** [comment]
[This `LIKE` query](https://github.com/drupal/drupal/blob/7.x/modules/user/user.module#L1273) should check for case-insensitive duplicates.

**Comment by DFurnes** [comment]
We ran through the checklist on Thor and things are working well! 🎉 There are now no "unsanitary numbers", only two(!!) duplicate mobiles (for accounts that didn't actually exist on production or Thor though, orphaned fields), and no duplicate emails.

I'm running through the `export-users-to-northstar` script on Thor now as a final spot check. So far we've done ~715,000 users and only ~2,000 have returned Northstar error codes – 37 from upsert conflicts (those accounts are now in delete queue to fix conflict); 1,685 because they have invalid email addresses, and 245 because the accounts no longer have an email or mobile (previously had an invalid mobile paired with an `@mobile` placeholder email).

Next steps:
- [x] Finish running the "export" script as a final spot check.
- [x] Update the `sanitize-mobile-numbers` & `unset-duplicate-mobiles` to provide a placeholder "invalid" email address when a user would be left with no mobile number or email. (#6985)
- [ ] Make a script to replace invalid email addresses with `bad-email-{uid}@dosomething.invalid` addresses to silence errors on those accounts. (Optional.)
- [x] Once "export" is done, run the `delete-bad-users-in-northstar` to fix up those remaining ~40ish conflicts that popped up on the export run-through.

🐎

**Comment by DFurnes** [comment]
Okay, so apparently the regex that PHP uses for `filter_var('...', FILTER_VALIDATE_EMAIL`)` is an [absolute _abomination_](https://github.com/php/php-src/blob/master/ext/filter/logical_filters.c#L607-L608) so maybe we'll just go and fix those as a step later on (based on the errors saved in the request log).

That being said I think everything's looking great to run on prod. All the entries I've spot-checked look great, the (very few) remaining conflicts are all correctly added to the delete queue, and the export script is almost exclusively 200s which is exactly what it should be.

**Comment by mshmsh5000** [comment]
[The Jamie Zawinski quote:](http://regex.info/blog/2006-09-15/247)

> Some people, when confronted with a problem, think “I know, I'll use regular expressions.”  Now they have two problems.

**Comment by DFurnes** [comment]
Currently in the process of running these on production (on `drupal.admin` in a tmux session).

There are logs of each script's output being stored in the `dosomething` user's home directory.

**Comment by DFurnes** [comment]
All three clean-up scripts have now completed successfully on production! 🎉 Kicking off the export script to run as a final check over the long weekend.
